### ISSUE TITLE
Question: Failure safety

### ISSUE THREAD

**Issue by kamiltoszek** [europeana/spring-batch-mongo#4]
I assume spring batch uses relation databases due to acid transactions. In case of interruption / failure from external sources, for example server shutting down, how safe is this mongo implementation? Will it rollback job state?

Is this a solution for https://github.com/spring-projects/spring-batch/issues/877 or temporary workaround?

**Comment by ikattey** [comment]
@KamilT This library should be considered a temporary solution, pending support for this functionality within Spring Batch itself.

None of the Mongo queries currently use transactions. This hasn't been an issue for us so far, but it's something we're considering adding to the library in future.

Does this answer your questions?

**Comment by gsergiu** [comment]
@kamiltoszek we have a specific workflow, shceduling individual jobs for each item, the item is written to the database only if the processing is successfull. So, we don't really need transactions and rollback. Failed jobs are written to the database. I will close the issue, as this doesn't apply to our usecase. In the meanwhile there is an official mongo support in spring-batch. So, we would recommend using the official version.

**Comment by fmbenhassine** [comment]
Hi,

I came across this repository with a search on Github. Congratulations to the folks from Europeana who worked on this!

We had this feature under our radar for a long time now, and I was working on adding support for MongoDB as Job repository in Spring Batch since last year. I wish I discovered this repository earlier. Looking at the code here, I believe we came to the same conclusion that there is a need for a persistence model that is different than the domain model. When I saw this approach being used in your repo, I was kind of assured that I am not completely out of track :-) I tried to summarize my findings on the matter here: https://github.com/spring-projects/spring-batch/issues/877#issuecomment-1783141473.

A notable difference with this repository is that I used a separate package for converters to/from persistence/domain models instead of using static methods in entities like in this repository. I believe this is required if we want the persistence model to be usable by other solutions (which is what we are intending to do in Spring Batch). In fact, MongoDB would be the first NoSQL solution supported by Spring Batch, and the preparatory work we have done paves the way for other NoSQL solutions.

The second notable difference is the transactionality of the job repository. I was able to use a `MongoTransactionManager` to proxy the base implementation provided by Spring Batch. And since MongoDB supports transactions since v4 IIRC, the job repository is now transactional. This should cover requirements like this: https://github.com/europeana/spring-batch-mongo/issues/4#issue-1269160663.

Anyway, since you certainly have experience running a Spring Batch job repository backed by MongoDB, I would be grateful if you give the [experimental feature](https://github.com/spring-projects-experimental/spring-batch-experimental?tab=readme-ov-file#mongodb-job-repository) a try and share your feedback with us. For the record, we are planning to include this feature in beta in the upcoming 5.2.

Best regards,
Mahmoud

**Comment by gsergiu** [comment]
Dear Ben, 
I'm glad to see that you are making steps to bring the spring-batch-mongo to official release.  We are quite bussy in this period, but I will try to provide some feedback next week, and eventually pinpoint some functionality we needed, and implemented in the projects shich depend on this shared code. One of the projects is the entity-management and here is the package with the extensions for spring batch: https://github.com/europeana/entity-management/tree/develop/entity-management-web/src/main/java/eu/europeana/entitymanagement/batch.

Maybe this class is a good starting point: https://github.com/europeana/entity-management/blob/develop/entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/config/JobLauncherConfig.java

Some of this requirements are:
- definition of different jobs with different configurations for the processors
- option to run synchronuously (for one or few items) or asynchronuously.

**Comment by gsergiu** [comment]
With regard to the use of Transactions, I'm not sure if that is the best apporach for our usecase because of the requirements we have for our usecase. 

Usecase: 
1. process large datasets (up to 500k items)
2. update entities/items with data from external data sources (temporary issues may occur)
3. validate data from external sources
4. merge data from different sources
5. save data to mongo and write to solr index for search

Requirements:
1. operate based on best effort approach, failures on individual items must not fail the whole job
2. retry the processing of items N times (i.e. 3) before the processing is declared to be failed

In short, we are processing the data in memory and write into the database only at the end if the data validation is successfull. Therefore, we don't have anything to rollback in the case that the validation fails.

**Comment by fmbenhassine** [comment]
Hi, 

Thank you for your feedback on this! Appreciated your time.

FYI, the feature has been merged in the main branch and will be available in Spring Batch 5.2.0.

Best regards,
Mahmoud
